util/saveimageutil/saveimageutil.go

```diff
@@ -0,0 +1,61 @@
+package saveimageutil
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/earthly/earthly/states/image"
+
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/pkg/errors"
+)
+
+// NewBuildImageSaver creates a new ImageSaver designed to be used within the buildkit Build function
+func NewBuildImageSaver(res *gwclient.Result) *ImageSaver {
+	return &ImageSaver{
+		addRef:  res.AddRef,
+		addMeta: res.AddMeta,
+	}
+}
+
+// NewExportImageSaver creates a new ImageSaver designed to be used to populate ref and metadata entries for the buildkit Export function
+func NewExportImageSaver(refs map[string]gwclient.Reference, metadata map[string][]byte) *ImageSaver {
+	return &ImageSaver{
+		addRef:  func(k string, ref gwclient.Reference) { refs[k] = ref },
+		addMeta: func(k string, v []byte) { metadata[k] = v },
+	}
+}
+
+// ImageSaver exposes a common interface for adding references and metadata to either buildkit's Build or Export functions
+type ImageSaver struct {
+	addRef  func(string, gwclient.Reference)
+	addMeta func(string, []byte)
+}
```

Comment:
Ideas for slightly improved name (s.t. the exporting of artifacts is possible in the future too):

* `RefsBuilder`
* `ResultBuilder`
* `GatewayResultBuilder`
* `GatewayExporterBuilder`